crates/re_renderer/src/error_handling/error_tracker.rs

```diff
@@ -0,0 +1,150 @@
+        let mut errors = self.errors.lock();
+        errors.retain(|_error, entry| {
+            // If the error was not logged on the just concluded frame, remove it.
+            device_timeline_frame_index == entry.last_occurred_frame_index
+        });
+    }
+
+    /// Handles an async error, calling [`ErrorTracker::handle_error`] as needed.
+    ///
+    /// `on_last_scope_resolved` is called when the last scope has resolved.
+    ///
+    /// `frame_index` should be the currently active frame index which is associated with the scope.
+    /// (by the time the scope finishes, the active frame index may have changed)
+    pub fn handle_error_future(
+        self: &std::sync::Arc<Self>,
+        error_scope_result: impl IntoIterator<
+            Item = impl std::future::Future<Output = Option<wgpu::Error>> + Send + 'static,
+        >,
+        frame_index: u64,
+        on_last_scope_resolved: impl Fn(&Self, u64) + Send + Sync + 'static,
+    ) {
+        let mut error_scope_result = error_scope_result.into_iter().peekable();
+        while let Some(error_future) = error_scope_result.next() {
+            if error_scope_result.peek().is_none() {
+                let err_tracker = self.clone();
+                handle_async_error(
+                    move |error| {
+                        if let Some(error) = error {
+                            err_tracker.handle_error(error, frame_index);
+                        }
+                        on_last_scope_resolved(&err_tracker, frame_index);
+                    },
+                    error_future,
+                );
+                break;
+            }
+
+            let err_tracker = self.clone();
+            handle_async_error(
+                move |error| {
+                    if let Some(error) = error {
+                        err_tracker.handle_error(error, frame_index);
+                    }
+                },
+                error_future,
+            );
+        }
+    }
+
+    /// Logs a wgpu error to the tracker.
+    ///
+    /// If the error happened already already, it will be deduplicated.
+    ///
+    /// `frame_index` should be the frame index associated with the error scope.
+    /// Since errors are reported on the `device timeline`, not the `content timeline`,
+    /// this may not be the currently active frame index!
+    pub fn handle_error(&self, error: wgpu::Error, frame_index: u64) {
+        match error {
+            wgpu::Error::OutOfMemory { source: _ } => {
+                re_log::error!("A wgpu operation caused out-of-memory: {error}");
+            }
+            wgpu::Error::Validation {
+                source: _source,
+                description,
+            } => {
+                let entry = ErrorEntry {
+                    last_occurred_frame_index: frame_index,
+                    description: description.clone(),
+                };
+
+                cfg_if::cfg_if! {
+                    if #[cfg(webgpu)] {
```

Comment:
Could this be a runtime `if` instead? (simpler code; the optimizer should remove the dead branch anyhow)

```suggestion
                if cfg!(webgpu) {
```

Reply:
sadly can't because one branch mentions wgpu_core types